src/lib/form-field/form-field.ts

```diff
@@ -344,10 +344,17 @@ export class MatFormField extends _MatFormFieldMixinBase
     return this._hasLabel() || this.appearance === 'legacy' && this._hasPlaceholder();
   }
 
-  /** Determines whether to display hints or errors. */
-  _getDisplayedMessages(): 'error' | 'hint' {
-    return (this._errorChildren && this._errorChildren.length > 0 &&
-        this._control.errorState) ? 'error' : 'hint';
+  /** Determines whether to display hints, errors or none. */
+  _getDisplayedMessages(): 'error' | 'hint' | 'none' {
+    return this.areErrorMessagesVisible() ? 'error' : this.areHintsVisible() ? 'hint' : 'none';
+  }
+
+  private areErrorMessagesVisible(): boolean {
+    return this._errorChildren && this._errorChildren.length > 0 && this._control.errorState;
```

Comment:
Move the `_control.errorState` in front so it can exit quicker if the control isn't in an error state:

```
return this._control.errorState && this._errorChildren && this._errorChildren.length > 0;
```